File: Chapter05/Java/StorageExample3/app/src/main/java/net/zenconsult/android/crypto/KeyManager.java
package net.zenconsult.android.crypto;

import android.content.Context;
import android.util.Log;

import java.io.ByteArrayOutputStream;
import java.io.FileInputStream;
import java.io.FileOutputStream;

public class KeyManager {

    private static final String TAG = "KeyManager";

    private static final String FILE1 = "id_value";
    private static final String FILE2 = "iv_value";

    private Context mContext;

    public KeyManager(Context context) {
        mContext = context;
    }

    public void setId(byte[] data) {
        writer(data, FILE1);
    }

    public void setIv(byte[] data) {
        writer(data, FILE2);
    }

    public byte[] getId() {
        return reader(FILE1);
    }

    public byte[] getIv() {
        return reader(FILE2);
    }

    private byte[] reader(String file) {
        byte[] data = null;
        FileInputStream fis = null;
        try {
            int bytesRead = 0;
            fis = mContext.openFileInput(file);
            ByteArrayOutputStream bos = new ByteArrayOutputStream();
            byte[] b = new byte[1024];
            while ((bytesRead = fis.read(b)) != -1) {
                bos.write(b, 0, bytesRead);
            }
            data = bos.toByteArray();
        } catch (Exception e) {
            Log.e(TAG, "reader=>error: ", e);
        } finally {
            if (fis != null) {
                try {
                    fis.close();
                } catch (Exception ignore) {}
            }
        }
        return data;
    }

    private void writer(byte[] data, String file) {
        FileOutputStream fos = null;
        try {
            fos = mContext.openFileOutput(file, Context.MODE_PRIVATE);
            fos.write(data);
            fos.flush();
        } catch (Exception e) {
            Log.e(TAG, "writer=>error: ", e);
        } finally {
            if (fos != null) {
                try {
                    fos.close();
                } catch (Exception ignore) {}
            }
        }
    }
}

File: Chapter05/Kotlin/StorageExample3/app/src/main/java/net/zenconsult/android/crypto/KeyManager.kt
package net.zenconsult.android.crypto

import android.content.Context
import java.io.ByteArrayOutputStream

class KeyManager(
    val mContext: Context
) {

    fun setId(data: ByteArray) {
        writer(data, FILE1)
    }

    fun setIv(data: ByteArray) {
        writer(data, FILE2)
    }

    fun getId(): ByteArray {
        return reader(FILE1)
    }

    fun getIv(): ByteArray {
        return reader(FILE2)
    }

    private fun writer(data: ByteArray, file: String) {
        val fos = mContext.openFileOutput(file, Context.MODE_PRIVATE)
        fos.write(data)
        fos.flush()
        fos.close()
    }

    private fun reader(file: String): ByteArray {
        val fis = mContext.openFileInput(file)
        val bos = ByteArrayOutputStream()
        val b = ByteArray(1024)
        var bytesRead = fis.read(b)
        while (bytesRead != -1) {
            bos.write(b, 0, bytesRead)
            bytesRead = fis.read(b)
        }
        return bos.toByteArray()
    }

    companion object {
        const val TAG = "KeyManager"
        const val FILE1 = "id_value"
        const val FILE2 = "iv_value"
    }
}